### PR TITLE
Issue 95: Use multipart upload API

### LINKED ISSUE #95: Use Multipart upload API to upload files larger than 5 GB
# Task Description
The multipart upload API is designed to improve the upload experience for larger objects. You can upload an object in parts. These object parts can be uploaded independently, in any order, and in parallel. You can use a multipart upload for objects from 5 MB to 5 TB in size.

# Tasks

The following tasks will need to be carried out:
* [ ] First create a multipart upload with `CreateMultipartUploadRequest` and get the upload id.
* [ ] Upload all the different parts of the object with the help of `UploadPartRequest` and `CompletedPart`.
* [ ] Finally call `completeMultipartUpload` operation with `CompletedMultipartUpload` request to tell S3 to merge all uploaded parts and finish the multipart operation.

# Task Relationships

This task:
* Is a follow-up of: #63

# Useful Links
* [Uploading objects](https://docs.aws.amazon.com/AmazonS3/latest/dev/UploadingObjects.html)
* [Amazon S3 multipart upload limits](https://docs.aws.amazon.com/AmazonS3/latest/dev/qfacts.html)
* [Upload objects in multiple parts](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/examples-s3-objects.html#list-objects)
* [Pull Request on Upplication/Amazon-S3-FileSystem-NIO2 (SDK v1)](https://github.com/Upplication/Amazon-S3-FileSystem-NIO2/pull/45)

# Help

* Our [chat channel](https://chat.carlspring.org/channel/s3fs-nio-community)
* Points of contact:
  * @carlspring
  * @steve-todorov
  * @sbespalov

### PR DESCRIPTION
**Pull request by ptirador**
# Pull Request Description

This pull request closes #95.

The multipart upload API has been integrated. It's is designed to improve the upload experience for larger objects. You can upload an object in parts. These object parts can be uploaded independently, in any order, and in parallel. You can use a multipart upload for objects from 5 MB to 5 TB in size (according to [Amazon S3 multipart upload limits](https://docs.aws.amazon.com/AmazonS3/latest/dev/qfacts.html)).
- `S3OutputStream` uses multipart uploads to transfer data to S3 without storing it on local disk first.
- `FileSystemProvider` returns this implementation for new `OutputStreams` instead of `S3SeekableByteChannel` if possible.

# Acceptance Test

* [x] Building the code with `mvn clean install -Pintegration-tests` still works.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [ ] Yes, please see [provide details here]
  * [x] No

### CODE REVIEW

File: src/main/java/org/carlspring/cloud/storage/s3fs/S3FileSystemProvider.java [hunk 1]
```diff
@@ -527,6 +533,77 @@ public InputStream newInputStream(Path path,
...
+            // remove irrelevant/ignored options
+            opts.remove(StandardOpenOption.WRITE);
+            opts.remove(StandardOpenOption.SPARSE);
+
+            if (!opts.isEmpty()) {
+                throw new UnsupportedOperationException(opts.iterator().next() + " not supported");
+            }
+
+            validateCreateAndTruncateOptions(path, s3Path, create, createNew, truncateExisting);
+        }
```
Comment (carlspring) [inline src/main/java/org/carlspring/cloud/storage/s3fs/S3FileSystemProvider.java:566]:
```suggestion
    public OutputStream newOutputStream(final Path path, final OpenOption... options)
            throws IOException
    {
        final S3Path s3Path = toS3Path(path);

        // validate options
        if (options.length > 0)
        {
            final Set<OpenOption> opts = new LinkedHashSet<>(Arrays.asList(options));

            // cannot handle APPEND here -> use newByteChannel() implementation
            if (opts.contains(StandardOpenOption.APPEND))
            {
                return super.newOutputStream(path, options);
            }

            if (opts.contains(StandardOpenOption.READ))
            {
                throw new IllegalArgumentException("READ not allowed");
            }

            final boolean create = opts.remove(StandardOpenOption.CREATE);
            final boolean createNew = opts.remove(StandardOpenOption.CREATE_NEW);
            final boolean truncateExisting = opts.remove(StandardOpenOption.TRUNCATE_EXISTING);

            // remove irrelevant/ignored options
            opts.remove(StandardOpenOption.WRITE);
            opts.remove(StandardOpenOption.SPARSE);

            if (!opts.isEmpty())
            {
                throw new UnsupportedOperationException(opts.iterator().next() + " not supported");
            }

            validateCreateAndTruncateOptions(path, s3Path, create, createNew, truncateExisting);
        }
```

File: src/main/java/org/carlspring/cloud/storage/s3fs/S3OutputStream.java [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+    private String getValueFromMetadata(final String key)
+    {
+        if (metadata.containsKey(key))
+        {
+            return metadata.get(key);
+        }
+
+        return null;
+    }
+}
```
Comment (carlspring) [inline src/main/java/org/carlspring/cloud/storage/s3fs/S3OutputStream.java:497]:
```suggestion
}

```

File: src/main/java/org/carlspring/cloud/storage/s3fs/S3FileSystemProvider.java (hunk 1 above)
Comment (ptirador) [reply src/main/java/org/carlspring/cloud/storage/s3fs/S3FileSystemProvider.java]: ¡Done!

File: src/main/java/org/carlspring/cloud/storage/s3fs/S3OutputStream.java (hunk 2 above)
Comment (ptirador) [reply src/main/java/org/carlspring/cloud/storage/s3fs/S3OutputStream.java]: Done!

File: src/test/java/org/carlspring/cloud/storage/s3fs/util/S3ClientMock.java [hunk 3]
```diff
@@ -166,6 +186,7 @@ void addFile(final Path bucket,
                  final byte[] content)
             throws IOException
     {
+
```
Comment (carlspring) [inline src/test/java/org/carlspring/cloud/storage/s3fs/util/S3ClientMock.java:189]:
```suggestion
```

**Review by carlspring — APPROVED** [review]

**Comment by carlspring** [comment]
Thanks for all your help! :)

**Pull request merged**